crates/matrix-sdk/src/sliding_sync/list/request_generator.rs

```diff
@@ -99,33 +102,163 @@ impl SlidingSyncListRequestGenerator {
+
+                // Create a new range, and use it as the current set of ranges.
+                self.ranges = vec![create_range(
+                    range_start,
+                    range_desired_size,
+                    maximum_number_of_rooms_to_fetch,
+                    maximum_number_of_rooms,
+                )?];
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Handle a sliding sync response, given a new maximum number of rooms.
+    pub(super) fn handle_response(
+        &mut self,
+        list_name: &str,
+        maximum_number_of_rooms: u32,
+    ) -> Result<SlidingSyncState, Error> {
         match &mut self.kind {
             SlidingSyncListRequestGeneratorKind::Paging {
                 number_of_fetched_rooms,
                 fully_loaded,
+                maximum_number_of_rooms_to_fetch,
                 ..
             }
             | SlidingSyncListRequestGeneratorKind::Growing {
                 number_of_fetched_rooms,
                 fully_loaded,
+                maximum_number_of_rooms_to_fetch,
                 ..
             } => {
-                *number_of_fetched_rooms = 0;
-                *fully_loaded = false;
+                let range_end: u32 = self.ranges.first().map(|r| *r.end()).ok_or_else(|| {
+                    Error::RequestGeneratorHasNotBeenInitialized(list_name.to_owned())
+                })?;
+
+                // Calculate the maximum bound for the range.
+                // At this step, the server has given us a maximum number of rooms for this
+                // list. That's our `range_maximum`.
+                let mut range_maximum = maximum_number_of_rooms;
+
+                // But maybe the user has defined a maximum number of rooms to fetch? In this
+                // case, let's take the minimum of the two.
+                if let Some(maximum_number_of_rooms_to_fetch) = maximum_number_of_rooms_to_fetch {
+                    range_maximum = min(range_maximum, *maximum_number_of_rooms_to_fetch);
+                }
+
+                // Finally, ranges are inclusive!
+                range_maximum = range_maximum.saturating_sub(1);
+
+                // Now, we know what the maximum bound for the range is.
+
+                // The current range hasn't reached its maximum, let's continue.
+                if range_end < range_maximum {
+                    // Update the number of fetched rooms forward. Do not forget that ranges are
+                    // inclusive, so let's add 1.
+                    *number_of_fetched_rooms = range_end.saturating_add(1);
+
+                    // The list is still not fully loaded.
+                    *fully_loaded = false;
+
+                    // Update the _list range_ to cover from 0 to `range_end`.
+                    // The list's range is different from the request generator (this) range.
+                    self.ranges = [0..=range_end].to_vec();
```

Comment:
I think we may have a problem here.

The `ranges` field is used in 2 ways. When a request is prepared, the `ranges` field is updated to contain the _ranges for the request_, for example:

* For Selective, it's always the same,
* For Growing, it's `0..=19`, `0..=39`, `0..=59` etc.
* For Paging, it's `0..=19`, `20..=39`, `40..=59` etc.

And this is how `Self::prepare_for_next_request` is computing things, because it doesn't use the `ranges` field: It only uses the `number_of_fetched_rooms` and `batch_size` (and other values to cap the ranges etc.). However, the new range is _stored_ in the `ranges` field.

Here, in `Self::handle_response`, we are storing ranges in the `ranges` field, but based on the received response. And, here what it does:

* For Selective, it's always the same,
* For Growing and for Paging, it's `0..=19, `0..=39`, `0..=59` etc.

Yeah, it always starts from 0!

Why are we storing the ranges in `ranges` in `Self::handle_response`? In `Self::prepare_for_next_request`, if the state is `FullyLoaded`, then we just return the value of the `ranges` field. It makes sense that it covers everything from the beginning, i.e. that it starts from 0.

All good right?

Nop.
Imagine that `SlidingSync` sends a first request. It prepares the request. A response is received, but it takes time to handle it. Meanwhile, a new request is sent. It can overwrite the `ranges` field. But `Self::handle_response` uses the `ranges` field to do its calculation. Maybe, maybe, it can create weird behaviors. I don't even want to think about those possible bugs.

Instead, I suggest that `Self::prepare_for_next_request` doesn't store the ranges into the `ranges` field, but instead _returns_ the new ranges. Thus, the method should be renamed `get_ranges_for_next_request` or `get_next_request_ranges`, whatever, but it must not store the ranges inside the `ranges` field. The `ranges` field must only contain the ranges for what we have received a valid response.

It's a small modification, but I think it will remove a lot of confusion (cf the comment of `Self::ranges` which states that the value has different meaning depending of when the method is called!), and a potential bug.

Reply:
That's very good, and the possible error you're talking about exists also in the current code, unless I'm mistaken.
Unfortunately, even in the next version, one piece of data is needed when handling the response (the end of the range, for growing/paginating generators). I'm storing this in the enum variant, but I suspect it should be stashed in the future so that there can't be any confusion between the request and the response handling (assuming the response will be handled in the right order too).